### PR TITLE
Implement locality aware load balancing

### PR DESCRIPTION
**Pull request by howardjohn**
Fixes https://github.com/istio/ztunnel/issues/862

### CODE REVIEW

File: proto/workload.proto
Comment (louiscryan) [inline proto/workload.proto:95]: typo?

File: proto/workload.proto
Comment (louiscryan) [inline proto/workload.proto:102]: no 'LOCALITY' value

File: proto/workload.proto
Comment (louiscryan) [inline proto/workload.proto:85]: Target is a pretty vague term, maybe 'Locality'

File: proto/workload.proto
Comment (louiscryan) [inline proto/workload.proto:86]: would ANY or NONE read more cleanly in docs.

"When the locality_preference is set to [NONE] traffic can be routed to any endpoint that is a member of the mesh"

File: proto/workload.proto
Comment (louiscryan) [inline proto/workload.proto:97]: Doc.  "Only if the set of endpoints in each target defined in routing_preference is empty or every endpoint is unreachable can an endpoint in the next target in routing_preference be used"

File: src/state.rs
Comment (ilrudie) [inline src/state.rs:259]: this seems to ignore failover/strict and always does a best match, yeah?

Comment (howardjohn) [reply src/state.rs:259]:
```
                        // Doesn't match all, and required to. Do not select this endpoint
                        if lb.mode == LoadBalancerMode::Strict && rank != lb.targets.len() {
                            return None;
                        }
```

should do it

File: src/state.rs
Comment (stevenctl) [inline src/state.rs:293]: We should document if we always expect the scopes to be "size-ordered"

Comment (howardjohn) [reply src/state.rs:293]: Its meant to be documented here:
```
   // In FAILOVER mode, endpoint selection will prefer endpoints that match all preferences, but failover to groups of endpoints
    // that match less (or, eventually, none) preferences.
    // For instance, with `[NETWORK, REGION, ZONE]`, we will send to:
    // 1. Endpoints matching `[NETWORK, REGION, ZONE]`
    // 2. Endpoints matching `[NETWORK, REGION]`
    // 3. Endpoints matching `[NETWORK]`
    // 4. Any endpoints
```

Which is also how failover works in Istio-envoy

File: src/state.rs
Comment (stevenctl) [inline src/state.rs:647]: just need to pass in src, right?

**Comment by stevenctl** [comment]
LGTM

File: src/state.rs
Comment (ilrudie) [reply src/state.rs:293]: I was also going to ask about this wrt the `break` being ok or not. I think it relies on the routing preferences being guaranteed ordered correctly.

File: src/state.rs
Comment (howardjohn) [reply src/state.rs:647]: ah good catch, forgot about this todo. Yep

File: src/state.rs
Comment (howardjohn) [reply src/state.rs:293]: My intended implementation was to make order matter here. Open to changes though

**Comment by howardjohn** [comment]
@stevenctl this should be good to go, will open istio side once this lands

File: src/state.rs
Comment (ilrudie) [reply src/state.rs:293]: I think that's fine as long as folks are aware that it matters.

**Review by stevenctl — APPROVED** [review]

File: proto/workload.proto
Comment (hzxuzhonghu) [inline proto/workload.proto:122]: With this api, how can user set it for `strict`? 

And how can we expand to support other algorithms with this struct?

Comment (daixiang0) [reply proto/workload.proto:122]: Package this as one algorithm if need to expand?

**Pull request merged**
